![Ironhack logo](https://i.imgur.com/1QgrNNw.png)

# Lab | Map, Reduce, Filter


## Introduction

In this lab, we will implement what we have learned about functional programming using the map, reduce, and filter functions. These functions allow us to pass an input and a transformation to a function and produce an output. 

## Getting Started

Follow the instructions and add your code and explanations as necessary. By the end of this lab, you will have learned about mapping, reducing, and filtering as well as applying functions in Pandas.


## Resources

[The Official Python Documentation on Mapping, Reducing, and Filtering](https://docs.python.org/3/howto/functional.html#built-in-functions)

[The `apply` Function in Pandas](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.apply.html)


# Before your start:
- Comment as much as you can
- Happy learning!

In [1]:
# import reduce from functools, numpy and pandas
import pandas as pd
import numpy as np
from functools import reduce

#### Fun fact: 
In 2005, the Python's language author Guido van van Rossum expressed his critics about map(), filter(), and reduce(). He advocated agaist those functions, affirming that loops were more legible and easy to handle. The discussion reached a point of excluding reduce from build-in function, isolating reduce into the functools library.

Source: https://www.artima.com/weblogs/viewpost.jsp?thread=98196

# Challenge 1 - Mapping

#### We will use the map function to clean up some words in a book.

In the following cell, we will read a text file containing the book The Prophet by Kahlil Gibran.

In [2]:
# Define the directory location
location = '../data/58585-0.txt'

In [3]:
# open the file using the open()
with open(location, 'r', encoding="utf8") as f:
    prophet = f.read().split(' ')

#### Let's remove the first 568 words since they contain information about the book but are not part of the book itself. 

Do this by removing from `prophet` elements 0 through 567 of the list (you can also do this by keeping elements 568 through the last element).

In [4]:
# Let's start by checking the number of elements inside the list 'prophet'
print(len(prophet))

13637


* The length of all prophet elements is 13.637. If we drop the first 568 elements, we're expecting to create a new one with 13.069 elements.

In [5]:
len(prophet)-568

13069

#### __Method 1__: just to keep in mind the string slide method using [ ]
* Now let's slice the data from 568 until the index '13637', or the list length

In [6]:
# define the max list's size
max_size = 568

In [7]:
# starting with a the list object followed by a bracket
prophet_568 = prophet[
    # define the max size
    max_size:]

In [8]:
# the new object's len is 568 elements, as we have expected
len(prophet_568)

13069

* Recheck: the element indexed as 568 is 'PROPHET\n\n|Almustafa,'.
* And so, the new object should start with the same element 'PROPHET\n\n|Almustafa,'.

In [9]:
prophet[568]

'PROPHET\n\n|Almustafa,'

In [10]:
prophet_568[0]

'PROPHET\n\n|Almustafa,'

If you look through the words, you will find that many words have a reference attached to them. For example, let's look at words 1 through 10.

Expected output:

````python
            ['PROPHET\n\n|Almustafa,',
             'the{7}',
             'chosen',
             'and',
             'the\nbeloved,',
             'who',
             'was',
             'a',
             'dawn',
             'unto']

````

In [11]:
# Slice the string from index 1 to 10
prophet1_10 = prophet_568[0:10]

In [12]:
prophet1_10

['PROPHET\n\n|Almustafa,',
 'the{7}',
 'chosen',
 'and',
 'the\nbeloved,',
 'who',
 'was',
 'a',
 'dawn',
 'unto']

In [13]:
# the new element lenghts 9 elements
len(prophet1_10)

10

#### The next step is to create a function that will remove references. 

We will do this by splitting the string on the `{` character and keeping only the part before this character. Write your function below.

In [14]:
def reference(x):
    '''
    Input: A string
    Output: The string with references removed
    
    Example:
    Input: 'the{7}'
    Output: 'the'
    '''
    # Your code here:
    return x.split('{')[0]

Now that we have our function, use the `map()` function to apply this function to our book, The Prophet. Return the resulting list to a new list called `prophet_reference`

In [15]:
# Your code here:

prophet_reference = list(map(reference, prophet_568))

In [16]:
prophet_reference

['PROPHET\n\n|Almustafa,',
 'the',
 'chosen',
 'and',
 'the\nbeloved,',
 'who',
 'was',
 'a',
 'dawn',
 'unto',
 'his',
 'own\nday,',
 'had',
 'waited',
 'twelve',
 'years',
 'in',
 'the',
 'city\nof',
 'Orphalese',
 'for',
 'his',
 'ship',
 'that',
 'was',
 'to\nreturn',
 'and',
 'bear',
 'him',
 'back',
 'to',
 'the',
 'isle',
 'of\nhis',
 'birth.\n\nAnd',
 'in',
 'the',
 'twelfth',
 'year,',
 'on',
 'the',
 'seventh\nday',
 'of',
 'Ielool,',
 'the',
 'month',
 'of',
 'reaping,',
 'he\nclimbed',
 'the',
 'hill',
 'without',
 'the',
 'city',
 'walls\nand',
 'looked',
 'seaward;',
 'and',
 'he',
 'beheld',
 'his\nship',
 'coming',
 'with',
 'the',
 'mist.\n\nThen',
 'the',
 'gates',
 'of',
 'his',
 'heart',
 'were',
 'flung\nopen,',
 'and',
 'his',
 'joy',
 'flew',
 'far',
 'over',
 'the',
 'sea.\nAnd',
 'he',
 'closed',
 'his',
 'eyes',
 'and',
 'prayed',
 'in',
 'the\nsilences',
 'of',
 'his',
 'soul.\n\n*****\n\nBut',
 'as',
 'he',
 'descended',
 'the',
 'hill,',
 'a',
 'sadness\nca

Another thing you may have noticed is that some words contain a line break. Let's write a function to split those words. Our function will return the string split on the character `\n`. Write your function in the cell below.

In [17]:
def line_break(x):
    '''
    Input: A string
    Output: A list of strings split on the line break (\n) character
        
    Example:
    Input: 'the\nbeloved'
    Output: ['the', 'beloved']
    '''
    
    return x.split('\n')    
    

Apply the `line_break` function to the `prophet_reference` list. Name the new list `prophet_line`.

In [18]:
prophet_line = list(map(line_break, prophet_reference));
prophet_line[0:10]

[['PROPHET', '', '|Almustafa,'],
 ['the'],
 ['chosen'],
 ['and'],
 ['the', 'beloved,'],
 ['who'],
 ['was'],
 ['a'],
 ['dawn'],
 ['unto']]

If you look at the elements of `prophet_line`, you will see that the function returned lists and not strings. Our list is now a list of lists. Flatten the list using list comprehension. Assign this new list to `prophet_flat`.

In [19]:
prophet_flat = [y for x in prophet_line for y in x if y != ''];
prophet_flat[0:10]

['PROPHET',
 '|Almustafa,',
 'the',
 'chosen',
 'and',
 'the',
 'beloved,',
 'who',
 'was',
 'a']

# Challenge 2 - Filtering

When printing out a few words from the book, we see that there are words that we may not want to keep if we choose to analyze the corpus of text. Below is a list of words that we would like to get rid of. Create a function that will return false if it contains a word from the list of words specified and true otherwise.

In [20]:
def word_filter(x):
    '''
    Input: A string
    Output: true if the word is not in the specified list and false if the word is in the list
        
    Example:
    word list = ['and', 'the']
    Input: 'and'
    Output: False
    
    Input: 'John'
    Output: True
    '''    
    word_list = ['and', 'the', 'a', 'an'];
    return x not in word_list;    
    

Use the `filter()` function to filter out the words speficied in the `word_filter()` function. Store the filtered list in the variable `prophet_filter`.

In [21]:
prophet_filter = list(filter(word_filter, prophet_flat));
prophet_filter[0:10]

['PROPHET',
 '|Almustafa,',
 'chosen',
 'beloved,',
 'who',
 'was',
 'dawn',
 'unto',
 'his',
 'own']

# Bonus Challenge

Rewrite the `word_filter` function above to not be case sensitive.

In [22]:
def word_filter_case(x):
   
    word_list = ['and', 'the', 'a', 'an']
    return x.lower() not in word_list 
    

In [23]:
prophet_filter_case = list(filter(word_filter_case, prophet_flat));
print("No case sensitive:", prophet_filter[30:40])
print('---------------------------')
print("Case sensitive:", prophet_filter_case[30:40])


No case sensitive: ['isle', 'of', 'his', 'birth.', 'And', 'in', 'twelfth', 'year,', 'on', 'seventh']
---------------------------
Case sensitive: ['isle', 'of', 'his', 'birth.', 'in', 'twelfth', 'year,', 'on', 'seventh', 'day']


# Challenge 3 - Reducing

#### Now that we have significantly cleaned up our text corpus, let's use the `reduce()` function to put the words back together into one long string separated by spaces. 

We will start by writing a function that takes two strings and concatenates them together with a space between the two strings.

In [24]:
def concat_space(a, b):
    '''
    Input:Two strings
    Output: A single string separated by a space
        
    Example:
    Input: 'John', 'Smith'
    Output: 'John Smith'
    '''
    return "{} {}".format(a,b);


In [25]:
# other method

def concat_space(a, b):
    '''
    Input:Two strings
    Output: A single string separated by a space
        
    Example:
    Input: 'John', 'Smith'
    Output: 'John Smith'
    '''
    return a + ' ' + b


Use the function above to reduce the text corpus in the list `prophet_filter` into a single string. Assign this new string to the variable `prophet_string`.

In [26]:
prophet_string = reduce(concat_space, prophet_filter)
prophet_string[0:1000]

'PROPHET |Almustafa, chosen beloved, who was dawn unto his own day, had waited twelve years in city of Orphalese for his ship that was to return bear him back to isle of his birth. And in twelfth year, on seventh day of Ielool, month of reaping, he climbed hill without city walls looked seaward; he beheld his ship coming with mist. Then gates of his heart were flung open, his joy flew far over sea. And he closed his eyes prayed in silences of his soul. ***** But as he descended hill, sadness came upon him, he thought in his heart: How shall I go in peace without sorrow? Nay, not without wound in spirit shall I leave this city. days of pain I have spent within its walls, long were nights of aloneness; who can depart from his pain his aloneness without regret? Too many fragments of spirit have I scattered in these streets, too many are children of my longing that walk naked among these hills, I cannot withdraw from them without burden ache. It is not garment I cast off this day, but skin

# Challenge 4 - Applying Functions to DataFrames

#### Our next step is to use the apply function to a dataframe and transform all cells.

To do this, we will load a dataset below and then write a function that will perform the transformation.

In [27]:
# Run this code:

# The dataset below contains information about pollution from PM2.5 particles in Beijing 

url = "https://archive.ics.uci.edu/ml/machine-learning-databases/00381/PRSA_data_2010.1.1-2014.12.31.csv"
pm25 = pd.read_csv(url)

Let's look at the data using the `head()` function.

Expected output:

>
>|    |   No |   year |   month |   day |   hour |   pm2.5 |   DEWP |   TEMP |   PRES | cbwd   |   Iws |   Is |   Ir |
|---:|-----:|-------:|--------:|------:|-------:|--------:|-------:|-------:|-------:|:-------|------:|-----:|-----:|
|  0 |    1 |   2010 |       1 |     1 |      0 |     nan |    -21 |    -11 |   1021 | NW     |  1.79 |    0 |    0 |
|  1 |    2 |   2010 |       1 |     1 |      1 |     nan |    -21 |    -12 |   1020 | NW     |  4.92 |    0 |    0 |
|  2 |    3 |   2010 |       1 |     1 |      2 |     nan |    -21 |    -11 |   1019 | NW     |  6.71 |    0 |    0 |
|  3 |    4 |   2010 |       1 |     1 |      3 |     nan |    -21 |    -14 |   1019 | NW     |  9.84 |    0 |    0 |
|  4 |    5 |   2010 |       1 |     1 |      4 |     nan |    -20 |    -12 |   1018 | NW     | 12.97 |    0 |    0 |

In [28]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/00381/PRSA_data_2010.1.1-2014.12.31.csv"

In [29]:
pm25 = pd.read_csv(url)

In [30]:
pm25.head()

No  year  month  day  hour  pm2.5  DEWP  TEMP    PRES cbwd    Iws  Is  Ir
0   1  2010      1    1     0    NaN   -21 -11.0  1021.0   NW   1.79   0   0
1   2  2010      1    1     1    NaN   -21 -12.0  1020.0   NW   4.92   0   0
2   3  2010      1    1     2    NaN   -21 -11.0  1019.0   NW   6.71   0   0
3   4  2010      1    1     3    NaN   -21 -14.0  1019.0   NW   9.84   0   0
4   5  2010      1    1     4    NaN   -20 -12.0  1018.0   NW  12.97   0   0

The next step is to create a function that divides a cell by 24 to produce an hourly figure. Write the function below.

In [31]:
def hourly(x):
    '''
    Input: A numerical value
    Output: The value divided by 24
        
    Example:
    Input: 48
    Output: 2.0
    '''
    return(x/24)    

Apply this function to the columns `Iws`, `Is`, and `Ir`. Store this new dataframe in the variable `pm25_hourly`.

In [32]:
pm25_hourly = pm25[['Iws','Is','Ir']].apply(hourly)
pm25_hourly.head()

Iws   Is   Ir
0  0.074583  0.0  0.0
1  0.205000  0.0  0.0
2  0.279583  0.0  0.0
3  0.410000  0.0  0.0
4  0.540417  0.0  0.0

#### Our last challenge will be to create an aggregate function and apply it to a select group of columns in our dataframe.

Write a function that returns the standard deviation of a column divided by the length of a column minus 1. Since we are using pandas, do not use the `len()` function. One alternative is to use `count()`. Also, use the numpy version of standard deviation.

In [33]:
def sample_sd(x):
    '''
    Input: A Pandas series of values
    Output: the standard deviation divided by the number of elements in the series
        
    Example:
    Input: pd.Series([1,2,3,4])
    Output: 0.3726779962
    '''
    
    # Your code here:
    return np.std(x) / (x.count() - 1);
    

In [34]:
pm25_hourly_sd = pm25[['Iws','Is','Ir']].apply(sample_sd);
pm25_hourly_sd.head()

Iws    0.001141
Is     0.000017
Ir     0.000032
dtype: float64